### PR TITLE
[SYCL][Bindless][Exp] Add Support For Unsampled Image Arrays

### PR DESCRIPTION
**Pull request by isaacault**
- Creation / destruction of unsampled image arrays
- Fetching / writing of unsampled image arrays
- `sycl::ext::oneapi::experimental::image_type::array` enum value added
- `sycl::ext::oneapi::experimental::image_descriptor::array_size` member added
- `sycl::ext::oneapi::experimental::image_descriptor::verify()` member function added

Correlated UR PR: [[Bindless][Exp] Add Support For Image Arrays #1274](https://github.com/oneapi-src/unified-runtime/pull/1274)

### CODE REVIEW

File: sycl/test-e2e/bindless_images/bindless_helpers.hpp
Comment (cppchedy) [inline sycl/test-e2e/bindless_images/bindless_helpers.hpp:44]: The vulkan interop tests also use these functions. can you remove them from the interop test and keep these in a single place? `add_kernel`, `fill_rand`, `printString`

File: sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp
Comment (cppchedy) [inline sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:77]: I think `run_ndim_test` can be fused into a single function that handles both 2D and 3D cases. you can take a look at https://github.com/intel/llvm/blob/ae24111fc96146b7daa8aea6d09325ddee76c927/sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp#L144

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (cppchedy) [inline sycl/include/sycl/ext/oneapi/bindless_images.hpp:1055]:
```suggestion
  detail::assert_unsampled_coords<CoordT>();
  constexpr size_t coordSize = detail::coord_size<CoordT>();
```

Comment (isaacault) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp:1056]: Good catch. Done.

File: sycl/test-e2e/bindless_images/bindless_helpers.hpp
Comment (isaacault) [reply sycl/test-e2e/bindless_images/bindless_helpers.hpp:45]: Using these functions here but leaving `printString` as is for now as it's not included in this file.

File: sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp
Comment (isaacault) [reply sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:17]: Good shout, done.

File: sycl/test-e2e/bindless_images/bindless_helpers.hpp
Comment (isaacault) [reply sycl/test-e2e/bindless_images/bindless_helpers.hpp]: Changed.

**Review by cppchedy — APPROVED** [review]

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (jchlanda) [inline libclc/ptx-nvidiacl/libspirv/images/image.cl:259]: This doesn't look right, casting between `int4` and `int4'? Did you mean:
```cpp
_DEFINE_VEC4_CAST(uint, int)
```

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (jchlanda) [inline libclc/ptx-nvidiacl/libspirv/images/image.cl:266]: Same here, `short4` to `short4`?

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (jchlanda) [inline libclc/ptx-nvidiacl/libspirv/images/image.cl:278]: Again, same type on both sides.

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (jchlanda) [inline libclc/ptx-nvidiacl/libspirv/images/image.cl:281]: Again.

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (jchlanda) [inline libclc/ptx-nvidiacl/libspirv/images/image.cl:284]: And again. A whole bunch more of the same type on both sides, going to ignore them from now on, but if they are indeed not needed, please address all the cases.

Seems that this patch adds some more casts, as well as reshuffles the existing ones. In an ideal git scenario, the reshuffling would be in a separate NFC PR, so it's easier to spot what is actually going on.

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (jchlanda) [inline libclc/ptx-nvidiacl/libspirv/images/image.cl:2639]: With 4 cases of `__nvvm_suld_[ARRAY_D]_array_[TYPE]_clamp_s_helper` I'd be tempted to wrap it in a macro, such is libclc style anyway 😄

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (jchlanda) [inline libclc/ptx-nvidiacl/libspirv/images/image.cl:2691]: Same here, maybe we could wrap cases for `int`, `short` and `char`in their own macros, seems to be a common pattern in each of the cases.

**Review by jchlanda — COMMENTED** [review]
I've only gone through libclc changes.

**Comment by isaacault** [comment]
@jchlanda My apologies, I accidentally rebased. I believe I've corrected my mistake, and you can see my changes addressing your feedback in [6b2ca14](https://github.com/intel/llvm/pull/12464/commits/6b2ca14b2793cf606790296a46318e2ac3ab3516).

File: sycl/source/detail/bindless_images.cpp
Comment (aelovikov-intel) [inline sycl/source/detail/bindless_images.cpp:36]: Is it missing support or they simply can't be per the specification? If the latter, why can't we rely on `verify` to throw instead? Same in `handler.cpp`.

File: sycl/source/handler.cpp
Comment (aelovikov-intel) [inline sycl/source/handler.cpp:1037]: How is this different from line 969-896? Can we outline into a helper to avoid copy-paste/make distinctins (if any) clearer? Same in multiple places below.

File: sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:2]: 1) How is this specific to linux? I don't think we have windows+cuda yet at all.
2) nit, it's possible to `REQUIRES: linux, cuda`

File: sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:4]:
```suggestion
// RUN: %{build} -o %t.out
```

File: sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:5]:
```suggestion
// RUN: %{run} %t.out
```

File: sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:16]:
```suggestion
#define VERBOSE_PRINT 0
```

and update the checks accordingly. Or (even better) switch to `inline constexpr bool` and `if constexpr`.

File: sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:75]: I think metaprogramming magic can unify these two versions neatly, but I won't insist.

File: sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:132]: I'd add a `static_assert` against `double` that would otherwise require `aspect:fp64`.

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (jchlanda) [reply libclc/ptx-nvidiacl/libspirv/images/image.cl]: Right, I understand now. Would you be bale to verify that the identity cast is indeed optimised out in the compiler and does not make its way to PTX?

**Comment by jchlanda** [comment]
> @jchlanda My apologies, I accidentally rebased. I believe I've corrected my mistake, and you can see my changes addressing your feedback in [6b2ca14](https://github.com/intel/llvm/pull/12464/commits/6b2ca14b2793cf606790296a46318e2ac3ab3516).

Not a problem, I've left one more question, otherwise the libclc part looks good. Thank you.

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (aarongreig) [reply sycl/plugins/unified_runtime/CMakeLists.txt:65]: https://github.com/orgs/intel/teams/unified-runtime-reviewers will not approve until this is updated, but we do have a policy in our repo to get approvals from other reviewers here before contributions there can proceed. We take full responsibility for blocking this kind of change

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (jchlanda) [reply libclc/ptx-nvidiacl/libspirv/images/image.cl]: I guess it's your call, so I'm happy to go with your judgement, but it would appear to me that for the casts in question they are either:
* optimised out, so there is no problem,
* or are generating the cast instruction in which case the macro should be re-factored?

**Review by aelovikov-intel — APPROVED** [review]
SYCL RT changes are good, cursory reading of e2e tests is fine stylistically, I'd expect the logic/completeness of the tests to be checked by images domain experts from Codeplay.

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (ldrumm) [reply libclc/ptx-nvidiacl/libspirv/images/image.cl]: This is the sort of work that never gets done, and there's no point refactoring if you have the opportunity to do it right the first time. If you can verify that the cast doesn't end up with extraneous code in the kernel now, I think that would placate Jakub somewhat. It might "just work" ;)

Comment (isaacault) [reply libclc/ptx-nvidiacl/libspirv/images/image.cl]: @jchlanda @ldrumm I understand your concerns.

I've looked into this and the output of `cuobjdump` called on my below e2e test is identical regardless of using these macros with the identity cast or an older version of `image.cl` where I don't have this redundancy.

With this, **I believe the identity cast is optimized out.**

Example PTX for the instance of reading 2 floats then writing their sum, where the results read would be cast with `cast_int_to_int()`.

```
suld.b.a1d.b32.clamp {%r10}, [%rd1, {%r8, %r9}];
mov.b32 %f1, %r10;                              
suld.b.a1d.b32.clamp {%r11}, [%rd2, {%r8, %r9}];
mov.b32 %f2, %r11;                              
add.rn.f32 %f3, %f1, %f2;                       
mov.b32 %r12, %f3;                              
sust.b.a1d.b32.clamp [%rd3, {%r8, %r9}], {%r12};
```

Note: This is my first time emitting and analysing PTX from one of the e2e tests, so for sanity check I removed the `as_float()` cast (which is used) and successfully observed this difference in the PTX emitted. That is to say, if you think my conclusion above isn't sufficient then please do share.

Comment (jchlanda) [reply libclc/ptx-nvidiacl/libspirv/images/image.cl]: Great, thank you for checking that, much appreciated.

**Review by jchlanda — APPROVED** [review]
libclc 👍

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:65]:
```suggestion
  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
  # commit 91c6068bb581acf72f02f1284ab0464ec4f03e34
  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
  # Date:   Fri Feb 23 13:39:03 2024 +0000
  #     Merge pull request #1274 from isaacault/iault/image_array_copy
  #     [Bindless][Exp] Add Support For Image Arrays
  set(UNIFIED_RUNTIME_TAG 91c6068bb581acf72f02f1284ab0464ec4f03e34)
```

**Review by kbenzie — COMMENTED** [review]
https://github.com/oneapi-src/unified-runtime/pull/1274 has been merged, please pull in the latest sycl branch changes and update the UR tag as suggested.

**Review by kbenzie — APPROVED** [review]

**Comment by isaacault** [comment]
@intel/llvm-gatekeepers Can this get merged? CI failure unrelated & already reported in https://github.com/intel/llvm/issues/12797 and https://github.com/intel/llvm/issues/12798

**Comment by kbenzie** [comment]
@intel/llvm-gatekeepers please merge, the Windows failure is being tracked in #12798

**Pull request merged**
